main/ZactuatorPWM.ino

```diff
@@ -0,0 +1,282 @@
+  // associated output pins.
+
+  // PWM outputs vary the light intensity linearly when used to control
+  // LEDs. But our eyes don't perceive actual linear changes as linear.
+  // This manifests as a problem when trying to have fine control
+  // over LEDs at very low levels.
+  // Using an 8-bit duty cycle for example only allows for 256 different
+  // linear levels of brightness.  Perceptually, the difference in
+  // brightness between levels 1 and 2 is very large - so to get fine
+  // control at dark levels, it's important that we use as high 
+  // resolution PWM as we can.
+  for(int i = 0; i < kNumChannels; ++i)
+  {
+#if ESP32
+    // I think this is the fastest frequency that allows for a 16-bit
+    // duty cycle on an ESP32
+    ledcSetup(i, 625.0, kNumDutyCycleBits);
+    ledcAttachPin(channelPins[i], i);
+#else
+#   error TODO: Write code here for different microcontrollers
+#endif
+    calibrationMinLinear[i] = 0.f;
+    calibrationMaxLinear[i] = 1.f;
+    calibrationGamma[i] = PWM_DEFAULT_GAMMA;
+  }
+
+  Log.trace(F("ZactuatorPWM setup done " CR));
+}
+
+// This applies a power curve to the input to try to make the inputs
+// be 'perceptually linear', as opposed to actually linear.
+static float perceptualToLinear(float perceptual, int channelIdx)
+{
+  return pow(perceptual, calibrationGamma[channelIdx]);
+}
+
+// If we're currently fading between states, then update those states
+void PWMLoop()
+{
+  previousUpdateTime = currentUpdateTime;
+  currentUpdateTime = millis();
+
+  if(fadeIsComplete)
+  {
+    return;
+  }
+
+  fadeIsComplete = true; //< If any channel isn't finished, we'll set this to false
+  for(int i = 0; i < kNumChannels; ++i)
+  {
+    // Calculate our lerp value through the current fade
+    long totalFadeDuration = fadeEndUpdateTime[i] - fadeStartUpdateTime[i];
+    float fadeLerpValue = 1.f;
+    if(totalFadeDuration > 0)
+    {
+      fadeLerpValue = (float) (currentUpdateTime - fadeStartUpdateTime[i]) / (float) totalFadeDuration;
+    }
+    if(fadeLerpValue >= 1.f)
+    {
+      currentValues[i] = targetValues[i];
+    }
+    else
+    {
+      currentValues[i] = ((targetValues[i] - fadeStartValues[i]) * fadeLerpValue) + fadeStartValues[i];
+      fadeIsComplete = false;
+    }
+  }
+
+  // Now convert these perceptually linear values into actually linear values
+  // and set the appropriate duty cycle for the outputs.
+  for(int i = 0; i < kNumChannels; ++i)
+  {
+    float linear = perceptualToLinear(currentValues[i], i);
+
+    // We always treat zero as zero so that it's truly off, regardless of the calibration data.
+    if(linear > 0.f)
+    {
+      // Remap according to the calibration
+      linear = (linear * (calibrationMaxLinear[i] - calibrationMinLinear[i])) + calibrationMinLinear[i];
+    }
+
+    long dutyCycle = (long) (linear * kUNormToDutyCycle);
+#if ESP32
+    ledcWrite(i, dutyCycle);
+#else
+#   error TODO: Write code here for different microcontrollers
```

Comment:
would it be possible to adapt it for esp8266 easily?

Reply:
Yes, it should be possible.

I don't have an esp8266 to test on, but I can write some code and make sure it at least compiles.  Pretty sure it'll work first time, because stuff usually does, right? 😉 

Also - I'll order one.

Reply:
The module now compiles for esp8266 and Arduino.